### PR TITLE
Use radix sort to sort postings when index sorting is enabled.

### PR DESCRIPTION
**Pull request by jpountz**
This switches to LSBRadixSorter instead of TimSorter to sort postings whose index options are `DOCS`. On a synthetic benchmark this yielded barely any difference in the case when the index order is the same as the sort order, or reverse, but almost a 3x speedup for writing postings in the case when the index order is mostly random.

### CODE REVIEW

**Comment by jpountz** [comment]
Here is the synthetic benchmark that I used if someone is interested in reproducing:

```java
  enum Order {
    RANDOM,
    ASC,
    DESC;
  }

  public static void main(String[] args) throws IOException {
    Order order = Order.RANDOM;
    Directory dir = FSDirectory.open(Paths.get("/tmp/a"));
    IndexWriterConfig cfg = new IndexWriterConfig(null);
    cfg.setInfoStream(new PrintStreamInfoStream(System.out));
    cfg.setMaxBufferedDocs(100_000);
    cfg.setRAMBufferSizeMB(IndexWriterConfig.DISABLE_AUTO_FLUSH);
    cfg.setIndexSort(new Sort(LongField.newSortField("sort_field", false, SortedNumericSelector.Type.MIN)));
    IndexWriter w = new IndexWriter(dir, cfg);
    Document doc = new Document();
    LongField sortField = new LongField("sort_field", 0);
    doc.add(sortField);
    StringField stringField1 = new StringField("string_field", "", Store.NO);
    doc.add(stringField1);
    StringField stringField2 = new StringField("string_field", "", Store.NO);
    doc.add(stringField2);
    StringField stringField3 = new StringField("string_field", "", Store.NO);
    doc.add(stringField3);
    for (int i = 0; i < 5_000_000; ++i) {
      long sortValue = switch (order) {
      case RANDOM -> i % 15;
      case ASC -> i;
      case DESC -> -i;
      };
      sortField.setLongValue(sortValue);
      stringField1.setStringValue(Integer.toBinaryString(i % 10));
      stringField2.setStringValue(Integer.toBinaryString(i % 100));
      stringField3.setStringValue(Integer.toBinaryString(i % 1000));
      w.addDocument(doc);
    }
  }
```

And flush times for postings:

| | Main  | Patch |
| ------ | ------------- | ------------- |
| Index sort matches indexing order | 6 | 7  |
| Index sort is reverse indexing order | 7  | 8  |
| Random sort | 27 | 10 |

File: lucene/core/src/java/org/apache/lucene/index/FreqProxTermsWriter.java [hunk 1]
```diff
@@ -379,27 +272,24 @@ public int advance(final int target) throws IOException {
...
-      return docs[docIt];
+    public long cost() {
+      return upTo;
     }
 
-    /** Returns the wrapped {@link PostingsEnum}. */
-    PostingsEnum getWrapped() {
-      return in;
+    @Override
+    public int freq() throws IOException {
```
Comment (zhaih) [inline lucene/core/src/java/org/apache/lucene/index/FreqProxTermsWriter.java:289]: So we're removing `freq` support because no one is really using it?

Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/index/FreqProxTermsWriter.java:289]: With this change, fields that have frequencies are now handled by `SortingPostingsEnum` while `SortingDocsEnum` focuses on fields that only index docs.

**Review by zhaih — APPROVED** [review]
Thanks for explaining! LGTM

**Comment by jpountz** [comment]
I purposedly introduced a bug to see what would fail, and only high-level tests that check early query termination or dynamic pruning failed, so I introduced lower-level tests that make sure that postings get reordered correctly.

**Pull request merged**